### PR TITLE
#1459 implements a first version (reading) of lazy loading

### PR DESCRIPTION
**Pull request by martin-lysk**

### CODE REVIEW

**Comment by martin-lysk** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by felixhaeberle** [comment]
@martin-lysk can't wait for this improvement! 🤩

File: lix/source-code/client/src/helpers/withLazyFetching.ts [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+/**
+ * Wraps a nodeishFs implementation with a js proxy for detailed logging, debugging and transparently replacing the file access behaviour.
+ * advantage of using this approach is that the underlying fs can also be swapped to something like lightingfs seamlessly.
+ */
+export const withLazyFetching = (
+	fs: NodeishFilesystem,
+	dir: string,
+	gitdir: string,
+	ref: string,
```
Comment (janfjohannes) [inline lix/source-code/client/src/helpers/withLazyFetching.ts:18]: ref, dir, gitdir, oidToFilePaths not required for first version, lets simplify args

File: lix/source-code/client/src/helpers/withLazyFetching.ts [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+			fs: fs,
+			gitdir,
+			http: http,
+			oids: [oid],
+		})
+
+		return execute()
+	}
+
+	return new Proxy(fs, {
```
Comment (janfjohannes) [inline lix/source-code/client/src/helpers/withLazyFetching.ts:174]: split the proxy wrapper and the lazzyloading interceptor back

File: lix/source-code/client/src/isomorphic-git-forks/basename.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+export function basename(path: string) {
```
Comment (janfjohannes) [inline lix/source-code/client/src/isomorphic-git-forks/basename.ts:1]: // forked files:
// link to whre from incl version, why forked, what was patched

File: lix/source-code/client/package.json [hunk 4]
```diff
@@ -25,6 +25,7 @@
 		"@lix-js/fs": "*",
 		"@octokit/types": "11.1.0",
 		"isomorphic-git": "1.24.5",
+		"path-browserify": "0.0.1",
```
Comment (janfjohannes) [inline lix/source-code/client/package.json:28]: still used?

File: lix/source-code/client/src/openRepository.ts [hunk 5]
```diff
@@ -92,17 +107,49 @@ export async function openRepository(
 
 	await pending
 
-	// delay all fs and repo operations until the repo clone and checkout have finished, this is preparation for the lazy feature
-	function delayedAction({ execute }: { execute: () => any }) {
-		if (pending) {
-			return pending.then(execute)
-		}
+	const oidToFilePaths = {} as { [oid: string]: string[] }
+	const filePathToOid = {} as { [filePath: string]: string }
```
Comment (janfjohannes) [inline lix/source-code/client/src/openRepository.ts:111]: replace cached mapping and use either walk ad hoc, or something like raw.listFiles or similar

File: lix/source-code/client/src/isomorphic-git-forks/git-fetch-request-helpers.ts [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+ * this is because the prefixed hex 008c -> dec 140 and the leght of the line is 136 + 4 length characters. The only line that doesn't take the
+ * length characters into account is the flush line. this is signalled by 0000
+ */
+export function encodePkLine(line: string) {
+
+    const flushLine = ""
+
+    if (line === flushLine) {
+        // flush lines don't take the padded hex into account length = 0 (ignoring the leading 0000)
+        return Buffer.from(padHex(4, 0), 'utf8');
```
Comment (janfjohannes) [inline lix/source-code/client/src/isomorphic-git-forks/git-fetch-request-helpers.ts:24]: btw. not sure why you use Buffer here, iso git removed using buffers in its api, could you explain? the iso git http client also does not use the old node buffer plyfill.

**Review by janfjohannes — CHANGES_REQUESTED** [review]

File: lix/source-code/client/src/openRepository.ts [hunk 7]
```diff
@@ -126,63 +187,431 @@ export async function openRepository(
...
+			}
+			const ref = head
+			// }
+
+			const { fetchHead, fetchHeadDescription } = await isoFetch({
+				fs: rawFs,
+				// cache,
+				http: httpWithLazyInjection(http, {
+					noneBlobFilter: true,
+					filterRefList: { ref: args.branch },
```
Comment (janfjohannes) [inline lix/source-code/client/src/openRepository.ts:506]: cant we replace filterRefList: { ref: args.branch }, just with ref as argument?

**Review by janfjohannes — CHANGES_REQUESTED** [review]

File: lix/source-code/client/src/helpers/withLazyFetching.ts [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+			// first of all we check if it exists - if so just return it - don't manipulate the index here
+			try {
+				await fs.stat(filePath)
+				return execute()
+			} catch (e) {
+				// TODO #1459 filter only file does not exists exception and throw others instead of continuing here...
+			}
+
+			// TODO #1459 this list is currently injected on creation of the proxy - this means it is not reactive to changes. I guess we should check this on the fly and walk the index every time?
+			const fileOid = filePathToOid[gitFilePath]
```
Comment (janfjohannes) [inline lix/source-code/client/src/helpers/withLazyFetching.ts:80]: todo: only reqyired to ad hoc copy the walk function here.

**Comment by samuelstroschein** [comment]
@martin-lysk can this PR be closed as stale?

**Comment by samuelstroschein** [comment]
@janfjohannes can this be closed? @martin-lysk mentioned that you use this PR as a reference

**Comment by janfjohannes** [comment]
@samuelstroschein  yes, please leave this pr, we use this as base ref of martins proof of concept. i will close this when the lazy loading real implementation is done.

**Comment by janfjohannes** [comment]
🎉

**Pull request closed without merging**